### ISSUE TITLE
FortiClient sees Ruby VSC extension as a threat

### ISSUE THREAD

**Issue by MonsieurDart** [rubyide/vscode-ruby#432]
### Your environment

- `vscode-ruby` version: 0.22.0
- Ruby version: 2.5
- Ruby version manager (if any): rvm 1.29.3
- VS Code version: 1.31.1
- Operating System: macOS Mojave
- Using language server? n/a

### Expected behavior

No threat should be detected.

### Actual behavior

FortiClient malware detection warns about a possible threat related to vscode-ruby:

<img width="670" alt="capture d ecran 2019-02-20 a 15 47 36" src="https://user-images.githubusercontent.com/295603/53102019-2f9a0680-352b-11e9-8b12-f5132cd67328.png">

According to TrendMicro and FortiGuard, this is related to the NodeJS EventStream vulnerability:

https://www.trendmicro.com/vinfo/au/threat-encyclopedia/malware/coinminer.js.malbtc.aa
https://fortiguard.com/encyclopedia/virus/7951871
https://github.com/dominictarr/event-stream/issues/116#issuecomment-441746370

Microsoft does not list this VSC extension in the list of problematic extensions, so maybe this warning by FortiNet is a false positive:

https://code.visualstudio.com/blogs/2018/11/26/event-stream

But on the other hand, there are other extensions infected but not listed in the MS blog post, for example:

https://community.platformio.org/t/trendmicro-sees-vscode-pio-extension-as-threat/5763 

For what it worth, here is the version of my FortiClient DBs (updated today):

<img width="666" alt="capture d ecran 2019-02-20 a 16 02 22" src="https://user-images.githubusercontent.com/295603/53102064-450f3080-352b-11e9-9fd5-d0668f2b6500.png">

**Comment by wingrunr21** [comment]
I am guessing FortiClient is slightly behind. I just released 0.22.0 this morning and that version is the one that has the vscode dependency upgraded (see 1d5637b)

**Comment by Firefishy** [comment]
Sophos AV and Kaspersky are also detecting threats.

**Comment by wingrunr21** [comment]
Ok good to know. I checked the current package structures for the vulnerable `event-stream` stuff (via `https://github.com/dominictarr/event-stream/issues/116`. See https://github.com/dominictarr/event-stream/issues/116) and came up with the correct version. `event-stream` is only pulled in via the `vscode` package and that's totally up to date

**Comment by MonsieurDart** [comment]
With my limited knowledge of NPM and the vulnerability, we’ve checked and it seems that the event-stream version used (resolved by NPM in package.json.lock) is 3.3.4 (i.e. the fixed one).

Mathieu Godart

**Comment by Firefishy** [comment]
Here is the totalvirus scan of the VSIX Package: https://www.virustotal.com/#/file/94493f18b160bc9e9698a6c0f5ab313eebd58b10b7a8c736018e39424786a441/detection

**Comment by Firefishy** [comment]
I see now the version on vscode exchange has been rolled back to v0.21.2, here is the scan of it: https://www.virustotal.com/#/file/395d0730d5ec92ed632905a513d4ee0c338edc42524bf345e062d77cb226a1f5/detection

**Comment by wingrunr21** [comment]
Alright. I'll have to dig more. `npm audit` doesn't bring up anything of note other than some dev dependency stuff.

**Comment by wingrunr21** [comment]
I'm really confused here. There are no dependencies here that depended on anything directly related to the `event-stream` issue other than the `vscode` dependency itself. That was upgraded precisely because npm/GitHub flagged it as a security issue. It doesn't make sense that old versions didn't get flagged and this one does (unless this is a red herring).

All of the flagged issues on those VirusTotal links appear to be related to `event-stream`/`flatmap-stream` problems. It's just not in the NPM tree...

**Comment by wingrunr21** [comment]
Ok. I think this is the `vscode` dependency.

`vscode` did not upgrade their `gulp-untar` dependency and that dependency pulls in `~3.3.4` of `event-stream`. Looks like it's been reported to them multiple times (see Microsoft/vscode-extension-vscode#126 and Microsoft/vscode-extension-vscode#128). Does anyone know how these scanners go about detecting this vulnerability?

I have no idea why this didn't get flagged before. The bad versions of `event-stream` and `flatmap-stream` are definitely in the `package-lock.json` files.

**Comment by wingrunr21** [comment]
And to add another twist:

I build the `vsix` by hand on a new machine and uploaded it to VirusTotal:

https://www.virustotal.com/#/file/492021cd1781a3feaa88748af82ad97dc4155319435e46e372439d63ef8763fe

None of the original detectors found anything.

**Comment by MonsieurDart** [comment]
On my .extension I have the versions 0.22.0 and 0.21.0.

I don’t know if this helps, but let me know if I can help or run more tests. 

Maybe we can ask the anti virus vendors why they detect a vulnerability?

**Comment by wingrunr21** [comment]
Can you search the 0.22.0 directory for the vulnerable packages? The extension root, `client`, and `server` directories should all have `node_modules` directories. I looked at my local one and didn't find anything.

**Comment by wingrunr21** [comment]
GAH I found it. I had accidentally included a `zip` file in the extension for a `vsix` that I was using for testing. That `vsix` was generated prior to upgrading the `vscode` extension. There's a vulnerable version of the `event-stream` package in the `node_modules` directory. 🤦‍♂️ 

Thanks everyone for bringing this to my attention.

I will push `0.22.1` tonight with that removed and a few other bugfixes. In the mean time, good news is that that zip file isn't actually used or executed so there's no real threat.

**Comment by MonsieurDart** [comment]
Yeah! Amazing work! 😄

Thanks for your very swift reaction. 

Mathieu Godart

**Comment by dylansavage** [comment]
The version number on the marketplace shows 0.21.1 instead of 0.22.1
<img width="268" alt="image" src="https://user-images.githubusercontent.com/7925135/53157118-bfd25d00-35b8-11e9-9b1b-ee280c2f0f99.png">

**Comment by wingrunr21** [comment]
Can you all take a look at `v0.22.1` that's now in the Marketplace?

Jiangmin still detects a generic Trojan. I'm wondering if it is using a different heuristic or is looking at something different. However, everything else is green.

https://www.virustotal.com/#/file/eb867668137a967271a54f23974a29f2ae8359a1156751b1ea53035102098619/detection

**Comment by MonsieurDart** [comment]
Well done!

I’ll check that asap and let you know if a warning fires here.

Thanks again for all your hard work. 😊

**Comment by modernplumbing** [comment]
A couple colleagues and I got flagged on the vsix.zip file in 0.22.0, and we also got flagged on a mysterious index.min.js file in our Temp folder that appeared around the same time. This thread mentions that the vsix.zip isn't used or executed, but is there any way that the index.min.js file could have come from 0.22.0?

We all had to have our computers rebuilt, so we can't look at the rogue index.min.js file anymore.
